### PR TITLE
Added the github actions for running tests

### LINKED ISSUE #35: Setup database for integeration tests to run and GitHub actions
Currently we don't have a test database which can get created for running the integration tests and gets destroyed post the tests are done. So this needs to be setup

I tried doing a POC during 25th Nov 23 - 26th Nov 23
But the container wasn't getting started.

**ETA**
Will do a POC 3days
Implement the database for integration tests 2 days
Write integration tests  - 2 days
PR reviews - 2 days

ETA - 7th Dec

### PR DESCRIPTION
**Pull request by vikhyat187**
Date: 13th Dec

Developer Name: @vikhyat187 

----

## Issue Ticket Number

-   closes #35 

## Description

This PR has the Github actions for running the tests

**Documentation Updated?**

-   [ ] Yes
-   [x] No

_If your feature adds a new API, then documentation related to the feature like API-Contracts, and Data-models  must be updated._

**Breaking Changes**

-   [ ] Yes
-   [x] No

_If your feature introduces breaking changes or if something is missing, please mention the related issue tickets._

**Development Tested?**

-   [x] Yes
-   [ ] No

_Confirm whether the changes have been tested locally during development._

**Tested in Staging?**

-   [ ] Yes
-   [x] No

_Indicate whether the changes have been tested in the staging environment._

**Under Feature Flag**

-   [ ] Yes
-   [x] No

_Specify if the changes are currently under a feature flag._

**Database Changes**

-   [ ] Yes
-   [x] No

_Indicate whether the changes include modifications to the database._

### Screenshots

Attach any relevant screenshots, such as test coverage reports, before and after images, or other visual aids.

## Test Coverage

Provide information on the test coverage for the changes made in this PR. Include details on the types of tests conducted and their outcomes.

## Additional Notes

Include any additional notes, considerations, or explanations that might be helpful for reviewers.

### CODE REVIEW

**Comment by iamitprakash** [comment]
@vikhyat187  @prakashchoudhary07 

cache: 'maven' option can significantly improve build speed by reusing downloaded Maven dependencies. However, it's important to invalidate the cache when necessary, such as after significant dependency updates.

mvnw path: The script uses ../.././mvnw to run Maven. This relative path might not be reliable across different repository structures. Consider using a more robust approach, such as specifying the full path to mvnw or using the maven tool provided by the actions/setup-java action.
Additional suggestions:

Consider adding steps for running tests after the build.

File: skill-tree /.github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Setup Java 17
+        uses: actions/setup-java@v3
+        with:
+          java-version: 17
+          distribution: 'temurin'
+          cache: 'maven'
```
Comment (yesyash) [inline skill-tree /.github/workflows/ci.yml:19]: - Do we want to add an architecture (ex: x64)?

File: skill-tree /.github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+      - name: Build with Maven
+        run: ../.././mvnw verify --file ../../pom.xml
```
Comment (yesyash) [inline skill-tree /.github/workflows/ci.yml:22]: - does this also run tests as suggested in the title of this pr?

**Review by yesyash — COMMENTED** [review]
The pr title menti

File: skill-tree /.github/workflows/ci.yml (hunk 2 above)
Comment (vikhyat187) [reply skill-tree /.github/workflows/ci.yml:22]: Yes it does

Comment (vikhyat187) [reply skill-tree /.github/workflows/ci.yml:22]: @yesyash I wanted to understand if the relative folder path given here is a proper format? Have not written the action script before.

File: skill-tree /.github/workflows/ci.yml (hunk 1 above)
Comment (vikhyat187) [reply skill-tree /.github/workflows/ci.yml:19]: The GitHub actions runner by default uses x86 runner, for the ARM runner we will have to setup our own runner 
Ref: https://github.com/orgs/community/discussions/25319

File: skill-tree /.github/workflows/ci.yml (hunk 2 above)
Comment (yesyash) [reply skill-tree /.github/workflows/ci.yml:22]: - it looks good to me on a first glance, if you want to test the action locally you can use a tool called (act)[https://github.com/nektos/act]

File: skill-tree /.github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,22 @@
+name: CI Build
+
+on:
+  pull_request:
+    branches:
+      - '**'
```
Comment (prakashchoudhary07) [inline skill-tree /.github/workflows/ci.yml:6]: on all branches?

Why didn't the current action run?

Comment (vikhyat187) [reply skill-tree /.github/workflows/ci.yml:6]: Not sure need to check, when I tried running the actions from here https://github.com/Real-Dev-Squad/skill-tree-backend/actions/runs/7197799606/workflow, this is running the intial file which I wrote.
@prakashchoudhary07

Comment (vikhyat187) [reply skill-tree /.github/workflows/ci.yml]: Have changed this to develop and main

**Comment by vikhyat187** [comment]
> mvnw path: The script uses ../.././mvnw to run Maven. This relative path might not be reliable across different repository structures. Consider using a more robust approach, such as specifying the full path to mvnw or using the maven tool provided by the actions/setup-java action.

Have changed this with mvn 

> cache: 'maven' option can significantly improve build speed by reusing downloaded Maven dependencies. However, it's important to invalidate the cache when necessary, such as after significant dependency updates.

removed the cache as in case of dependency update, the cache value will not updated, and has to be invalidated manually periodically.
Link - https://hackernoon.com/working-with-github-actions-cache-github-actions-on-steroids

> Consider adding steps for running tests after the build.
verify command does compile, packaging and running integration tests as well
https://stackoverflow.com/questions/66876221/what-is-the-difference-between-mvn-verify-vs-mvn-test#:~:text=Based%20on%20this%2C%20the%20first%20point%20is%20that%20verify%20includes%20test.&text=TEST%20%2D%20test%20the%20compiled%20source,ensure%20quality%20criteria%20are%20met

**Review by Ajeyakrishna-k — APPROVED** [review]

**Review by vinayak-trivedi — APPROVED** [review]

**Pull request merged**